guides/source/testing.md

```diff
@@ -1242,378 +1375,320 @@ assert_response :redirect
-
+We were able to successfully test a very small workflow for visiting our blog
+and creating a new article. If we wanted to take this further we could add tests
+for adding comments, editing comments or removing articles. Integration tests
+are a great place to experiment with all kinds of use cases for our
+applications.
 
-Functional Tests for Your Controllers
--------------------------------------
+### Helpers Available for Integration Tests
 
-In Rails, testing the various actions of a controller is a form of writing functional tests. Remember your controllers handle the incoming web requests to your application and eventually respond with a rendered view. When writing functional tests, you are testing how your actions handle the requests and the expected result or response, in some cases an HTML view.
+Here are some of the other helpers we can choose from in our integration tests.
 
-### What to Include in Your Functional Tests
+For dealing with the integration test runner, see
+[`ActionDispatch::Integration::Runner`](https://api.rubyonrails.org/classes/ActionDispatch/Integration/Runner.html).
 
-You should test for things such as:
+When performing requests, we will have
+[`ActionDispatch::Integration::RequestHelpers`](https://api.rubyonrails.org/classes/ActionDispatch/Integration/RequestHelpers.html)
+available for our use.
 
-* was the web request successful?
-* was the user redirected to the right page?
-* was the user successfully authenticated?
-* was the appropriate message displayed to the user in the view?
-* was the correct information displayed in the response?
+If we need to upload files, take a look at
+[`ActionDispatch::TestProcess::FixtureFile`](https://api.rubyonrails.org/classes/ActionDispatch/TestProcess/FixtureFile.html)
+to help.
 
-The easiest way to see functional tests in action is to generate a controller using the scaffold generator:
+If we need to modify the session, or state of our integration test, take a look
+at
+[`ActionDispatch::Integration::Session`](https://api.rubyonrails.org/classes/ActionDispatch/Integration/Session.html)
+to help.
 
-```bash
-$ bin/rails generate scaffold_controller article title:string body:text
-...
-create  app/controllers/articles_controller.rb
-...
-invoke  test_unit
-create    test/controllers/articles_controller_test.rb
-...
-```
+System Testing
+--------------
 
-This will generate the controller code and tests for an `Article` resource.
-You can take a look at the file `articles_controller_test.rb` in the `test/controllers` directory.
+Similarly to integration testing, system testing also allows you to test how the
+components of your app work together, but from the point of view of a user. It
+does this by running tests in either a real or a headless browser (a browser
+which runs in the background without opening a visible window). System tests use
+Capybara under the hood.
```

Comment:
We should probably mention Capybara's assertions as well (assert_selector and assert_text are used in the examples).
Not sure if this would the right location. Maybe it would be better to mention them after they have been used?
```suggestion
Capybara under the hood. [Capybara-powered
Assertions](https://rubydoc.info/github/teamcapybara/capybara/master/Capybara/Minitest/Assertions)
```

Or maybe we should list some of the common ones in a table like we do for other assertions:
https://edgeguides.rubyonrails.org/testing.html#available-assertions